File: ts/src/merge.ts
import { EpubMergeError } from "./errors.js";
import { baseZipEntries, buildFlatNav, buildOpf, buildVolumeNav, MERGE_MANIFEST_PATH, readSourceBook } from "./epub-io.js";
import type { InputFile, ManifestItem, MergeOptions, MergeResult, SourceBook, TocEntry } from "./models.js";
import { dirnamePosix, relPosix, safeBasename, sha256, utf8Decode, utf8Encode } from "./utils.js";
import { writeZip, type ZipEntry } from "./zip.js";
import { sortSources } from "./ordering.js";

const IMAGE_FONT_TYPES = new Set([
  "font/ttf", "font/otf", "font/woff", "font/woff2",
  "application/font-sfnt", "application/font-woff", "application/vnd.ms-opentype",
  "application/x-font-ttf", "application/x-font-opentype", "application/x-font-truetype"
]);

export function mergeEpubs(outputName: string, inputs: InputFile[], options: MergeOptions = {}): MergeResult {
  const structure = options.structure ?? "volume";
  if (structure !== "volume" && structure !== "flat") throw new EpubMergeError("structure must be 'volume' or 'flat'");
  if (!inputs.length) throw new EpubMergeError("At least one input EPUB is required");

  let sources = inputs.map(readSourceBook);
  rejectDuplicateBasenames(sources);
  if (!options.inputOrder) sources = sortSources(sources);
  if (structure === "flat") rejectDuplicateFlatTitles(sources);

  const title = options.title ?? outputName.replace(/\.epub$/i, "");
  const zipEntries: ZipEntry[] = baseZipEntries();
  const outputItems: ManifestItem[] = [];
  const spineIds: string[] = [];
  const resourcePool = new Map<string, string>();
  const manifestHrefIds = new Map<string, string>();
  const written = new Set<string>();
  const volumeToc: Array<{ title: string; href: string; entries: TocEntry[] }> = [];
  const flatToc: TocEntry[] = [];
  const manifestSources: Record<string, unknown>[] = [];
  let firstHref = "#";

  sources.forEach((source, sourceIndex) => {
    const prefix = sourceIndex === 0 ? "" : `v${sourceIndex}/`;
    const hrefMap = new Map<string, string>();
    const idMap = new Map<string, string>();
    const fileRecords: Record<string, unknown>[] = [];
    const rewrites: Record<string, Record<string, string>> = {};

    for (const item of source.manifestItems) {
      if (item.properties.includes("nav") || item.mediaType === "application/x-dtbncx+xml") continue;
      const data = source.itemData.get(item.href);
      if (!data) throw new EpubMergeError(`${source.basename}: missing item data ${item.href}`);
      const preferred = `${prefix}${item.href}`;
      let finalHref = preferred;
      if (canPool(item.mediaType)) {
        const key = `${item.mediaType.toLowerCase()}:${sha256(data)}`;
        finalHref = resourcePool.get(key) ?? preferred;
        resourcePool.set(key, finalHref);
      }
      hrefMap.set(item.href, finalHref);
      const newId = `s${sourceIndex}_${item.id}`;
      let ownerId = manifestHrefIds.get(finalHref);
      if (!ownerId) {
        ownerId = newId;
        manifestHrefIds.set(finalHref, ownerId);
        outputItems.push({ id: ownerId, href: finalHref, mediaType: item.mediaType, properties: item.properties.filter((p) => p !== "nav") });
      }
      idMap.set(item.id, ownerId);
      fileRecords.push({
        id: item.id,
        href: item.href,
        media_type: item.mediaType,
        properties: item.properties,
        merged_href: finalHref
      });
    }

    for (const item of source.manifestItems) {
      if (item.properties.includes("nav") || item.mediaType === "application/x-dtbncx+xml") continue;
      let data = source.itemData.get(item.href);
      if (!data) throw new EpubMergeError(`${source.basename}: missing item data ${item.href}`);
      const rewriteMap = rewriteMapForItem(item, prefix, hrefMap);
      if (Object.keys(rewriteMap).length) {
        data = rewriteRefs(data, rewriteMap);
        rewrites[item.href] = rewriteMap;
      }
      const finalHref = hrefMap.get(item.href)!;
      const zipName = `OEBPS/${finalHref}`;
      if (!written.has(zipName)) {
        zipEntries.push({ name: zipName, data });
        written.add(zipName);
      }
    }

    for (const id of source.spineIds) spineIds.push(idMap.get(id)!);
    const remapped = source.toc.map((entry) => ({ title: entry.title, href: remapTocHref(entry.href, hrefMap) }));
    if (firstHref === "#" && remapped.length) firstHref = remapped[0].href;
    volumeToc.push({ title: source.title, href: remapped[0]?.href ?? "#", entries: remapped });
    flatToc.push(...remapped);
    manifestSources.push({
      basename: source.basename,
      sha256: source.sha256,
      title: source.title,
      language: source.language,
      creators: source.creators,
      opf_path: source.opfPath,
      files: fileRecords,
      spine: source.spineIds,
      toc: source.toc,
      rewrites
    });
  });

  zipEntries.push({ name: "OEBPS/nav-merged.xhtml", data: structure === "flat" ? buildFlatNav(title, firstHref, flatToc) : buildVolumeNav(title, firstHref, volumeToc) });
  zipEntries.push({ name: "OEBPS/content.opf", data: buildOpf(title, sources[0].language, [], outputItems, spineIds) });
  zipEntries.push({
    name: MERGE_MANIFEST_PATH,
    data: utf8Encode(JSON.stringify({
      schema: "epub-merge-tool/v1",
      tool_version: "ts-0.1.0",
      structure,
      title,
      language: sources[0].language,
      sources: manifestSources
    }, null, 2))
  });
  return { name: outputName, data: writeZip(zipEntries) };
}

function rejectDuplicateBasenames(sources: SourceBook[]): void {
  const seen = new Set<string>();
  for (const source of sources) {
    const basename = safeBasename(source.basename);
    if (seen.has(basename)) throw new EpubMergeError(`Duplicate source basename: ${basename}`);
    seen.add(basename);
  }
}

function rejectDuplicateFlatTitles(sources: SourceBook[]): void {
  const seen = new Map<string, string>();
  for (const source of sources) {
    for (const entry of source.toc) {
      const key = entry.title.trim();
      const other = seen.get(key);
      if (other) throw new EpubMergeError(`Duplicate flat chapter title '${entry.title}' in ${source.basename} and ${other}`);
      seen.set(key, source.basename);
    }
  }
}

function canPool(mediaType: string): boolean {
  const normalized = mediaType.toLowerCase();
  return normalized.startsWith("image/") || IMAGE_FONT_TYPES.has(normalized);
}

function rewriteMapForItem(item: ManifestItem, prefix: string, hrefMap: Map<string, string>): Record<string, string> {
  if (!item.mediaType.toLowerCase().endsWith("xhtml+xml")) return {};
  const chapterDir = dirnamePosix(`${prefix}${item.href}`);
  const mapping: Record<string, string> = {};
  for (const [original, finalHref] of hrefMap.entries()) {
    const preferred = `${prefix}${original}`;
    if (preferred === finalHref) continue;
    mapping[original] = relPosix(chapterDir, finalHref);
  }
  return mapping;
}

function rewriteRefs(data: Uint8Array, mapping: Record<string, string>): Uint8Array {
  let text = utf8Decode(data);
  for (const [oldHref, newHref] of Object.entries(mapping)) {
    text = text
      .replaceAll(`src="${oldHref}"`, `src="${newHref}"`)
      .replaceAll(`src='${oldHref}'`, `src='${newHref}'`)
      .replaceAll(`href="${oldHref}"`, `href="${newHref}"`)
      .replaceAll(`href='${oldHref}'`, `href='${newHref}'`);
  }
  return utf8Encode(text);
}

function remapTocHref(href: string, hrefMap: Map<string, string>): string {
  const [base, fragment] = href.split("#", 2);
  const mapped = hrefMap.get(base);
  if (!mapped) throw new EpubMergeError(`TOC href '${href}' does not reference a copied manifest item`);
  return fragment === undefined ? mapped : `${mapped}#${fragment}`;
}

File: src/epub_merge_tool/merge.py
from __future__ import annotations

import hashlib
import json
import posixpath
import zipfile
from pathlib import Path

from . import __version__
from .epub_io import (
    MERGE_MANIFEST_PATH,
    build_flat_nav_html,
    build_nav_html,
    build_opf,
    read_source_book,
    safe_basename,
    write_epub_container,
    write_mimetype_first,
)
from .errors import EpubMergeError
from .models import ManifestItem, SourceBook, TocEntry
from .ordering import sort_sources


IMAGE_OR_FONT_TYPES = {
    "font/ttf",
    "font/otf",
    "font/woff",
    "font/woff2",
    "application/font-sfnt",
    "application/font-woff",
    "application/vnd.ms-opentype",
    "application/x-font-ttf",
    "application/x-font-opentype",
    "application/x-font-truetype",
}


def merge_epubs(
    output_path: Path | str,
    input_paths: list[Path | str],
    *,
    title: str | None = None,
    structure: str = "volume",
    input_order: bool = False,
) -> Path:
    if structure not in {"volume", "flat"}:
        raise EpubMergeError("structure must be 'volume' or 'flat'")
    if not input_paths:
        raise EpubMergeError("At least one input EPUB is required")

    sources = [read_source_book(Path(path)) for path in input_paths]
    _reject_duplicate_basenames(sources)
    if not input_order:
        sources = sort_sources(sources)
    if structure == "flat":
        _reject_duplicate_flat_titles(sources)

    output = Path(output_path).expanduser()
    output.parent.mkdir(parents=True, exist_ok=True)
    book_title = title or output.stem
    language = sources[0].language if sources else "en"

    manifest_items: list[ManifestItem] = []
    spine_ids: list[str] = []
    manifest_sources: list[dict] = []
    resource_pool: dict[tuple[str, str], str] = {}
    manifest_href_ids: dict[str, str] = {}
    zip_written: set[str] = set()
    source_tocs: list[tuple[str, str, list[TocEntry]]] = []
    flat_toc: list[TocEntry] = []
    first_href = "#"

    with zipfile.ZipFile(output, "w", compression=zipfile.ZIP_DEFLATED) as out:
        write_mimetype_first(out)
        write_epub_container(out)

        for source_index, source in enumerate(sources):
            prefix = "" if source_index == 0 else f"v{source_index}/"
            href_map: dict[str, str] = {}
            id_map: dict[str, str] = {}
            file_records: list[dict] = []
            rewrites: dict[str, dict[str, str]] = {}

            for item in source.manifest_items:
                if "nav" in item.properties:
                    continue
                data = source.item_data[item.href]
                preferred_href = f"{prefix}{item.href}"
                output_href = preferred_href
                if _can_pool(item.media_type):
                    digest = hashlib.sha256(data).hexdigest()
                    output_href = resource_pool.setdefault((item.media_type.lower(), digest), preferred_href)
                href_map[item.href] = output_href
                new_id = f"s{source_index}_{item.item_id}"
                owner_id = manifest_href_ids.get(output_href)
                if owner_id is None:
                    manifest_href_ids[output_href] = new_id
                    owner_id = new_id
                    manifest_items.append(
                        ManifestItem(
                            item_id=new_id,
                            href=output_href,
                            media_type=item.media_type,
                            properties=tuple(prop for prop in item.properties if prop != "nav"),
                        )
                    )
                id_map[item.item_id] = owner_id
                file_records.append(
                    {
                        "id": item.item_id,
                        "href": item.href,
                        "media_type": item.media_type,
                        "properties": list(item.properties),
                        "merged_href": output_href,
                    }
                )

            for item in source.manifest_items:
                if "nav" in item.properties:
                    continue
                data = source.item_data[item.href]
                rewrite_map = _rewrite_map_for_item(item, prefix, href_map)
                if rewrite_map:
                    data = _rewrite_refs(data, rewrite_map)
                    rewrites[item.href] = rewrite_map
                output_href = href_map[item.href]
                zip_name = f"OEBPS/{output_href}"
                if zip_name not in zip_written:
                    out.writestr(zip_name, data, compress_type=zipfile.ZIP_DEFLATED)
                    zip_written.add(zip_name)

            for item_id in source.spine_ids:
                spine_ids.append(id_map[item_id])

            remapped_toc = [
                TocEntry(entry.title, _remap_toc_href(entry.href, href_map)) for entry in source.toc
            ]
            if first_href == "#" and remapped_toc:
                first_href = remapped_toc[0].href
            source_href = remapped_toc[0].href if remapped_toc else "#"
            source_tocs.append((source.title, source_href, remapped_toc))
            flat_toc.extend(remapped_toc)
            manifest_sources.append(
                {
                    "basename": source.basename,
                    "sha256": source.sha256,
                    "title": source.title,
                    "language": source.language,
                    "creators": list(source.creators),
                    "opf_path": source.opf_path,
                    "files": file_records,
                    "spine": list(source.spine_ids),
                    "toc": [{"title": entry.title, "href": entry.href} for entry in source.toc],
                    "rewrites": rewrites,
                }
            )

        if structure == "flat":
            nav = build_flat_nav_html(book_title, first_href, flat_toc)
        else:
            nav = build_nav_html(book_title, first_href, source_tocs)
        out.writestr("OEBPS/nav-merged.xhtml", nav, compress_type=zipfile.ZIP_DEFLATED)
        out.writestr(
            "OEBPS/content.opf",
            build_opf(book_title, language, (), manifest_items, spine_ids),
            compress_type=zipfile.ZIP_DEFLATED,
        )
        manifest = {
            "schema": "epub-merge-tool/v1",
            "tool_version": __version__,
            "structure": structure,
            "title": book_title,
            "language": language,
            "sources": manifest_sources,
        }
        out.writestr(
            MERGE_MANIFEST_PATH,
            json.dumps(manifest, ensure_ascii=False, indent=2).encode("utf-8"),
            compress_type=zipfile.ZIP_DEFLATED,
        )
    return output


def _reject_duplicate_basenames(sources: list[SourceBook]) -> None:
    seen: set[str] = set()
    for source in sources:
        basename = safe_basename(source.path)
        if basename in seen:
            raise EpubMergeError(f"Duplicate source basename: {basename}")
        seen.add(basename)


def _reject_duplicate_flat_titles(sources: list[SourceBook]) -> None:
    seen: dict[str, str] = {}
    for source in sources:
        for entry in source.toc:
            key = entry.title.strip()
            if key in seen:
                raise EpubMergeError(
                    f"Duplicate flat chapter title {entry.title!r} in {source.basename} and {seen[key]}"
                )
            seen[key] = source.basename


def _can_pool(media_type: str) -> bool:
    normalized = media_type.lower()
    return normalized.startswith("image/") or normalized in IMAGE_OR_FONT_TYPES


def _rewrite_map_for_item(item: ManifestItem, prefix: str, href_map: dict[str, str]) -> dict[str, str]:
    if not item.media_type.lower().endswith("xhtml+xml"):
        return {}
    mapping: dict[str, str] = {}
    chapter_dir = posixpath.dirname(f"{prefix}{item.href}")
    for original_href, final_href in href_map.items():
        preferred_href = f"{prefix}{original_href}"
        if final_href == preferred_href:
            continue
        relative = posixpath.relpath(final_href, chapter_dir or ".")
        mapping[original_href] = relative
    return mapping


def _rewrite_refs(data: bytes, mapping: dict[str, str]) -> bytes:
    text = data.decode("utf-8")
    for old, new in mapping.items():
        text = text.replace(f'src="{old}"', f'src="{new}"')
        text = text.replace(f"src='{old}'", f"src='{new}'")
        text = text.replace(f'href="{old}"', f'href="{new}"')
        text = text.replace(f"href='{old}'", f"href='{new}'")
    return text.encode("utf-8")


def _remap_toc_href(href: str, href_map: dict[str, str]) -> str:
    base, sep, fragment = href.partition("#")
    mapped = href_map.get(base)
    if mapped is None:
        raise EpubMergeError(f"TOC href {href!r} does not reference a copied manifest item")
    return f"{mapped}{sep}{fragment}" if sep else mapped
